connectors/src/connectors/notion/temporal/workflows/deletion_crawl.ts

```diff
@@ -0,0 +1,123 @@
+import {
+  condition,
+  continueAsNew,
+  proxyActivities,
+  setHandler,
+  workflowInfo,
+} from "@temporalio/workflow";
+
+import type * as activities from "@connectors/connectors/notion/temporal/activities";
+import type { NotionDeletionCrawlSignal } from "@connectors/connectors/notion/temporal/signals";
+import { notionDeletionCrawlSignal } from "@connectors/connectors/notion/temporal/signals";
+import type { ModelId } from "@connectors/types";
+
+const {
+  checkResourceAndQueueRelated,
+  batchDeleteResources,
+  clearWorkflowCache,
+} = proxyActivities<typeof activities>({
+  startToCloseTimeout: "10 minute",
+});
+
+const DELETE_BATCH_SIZE = 100;
+
+/**
+ * Signal-based deletion crawl workflow.
+ * Receives signals for resources to check, maintains a "seen" cache to avoid duplicates,
+ * and recursively checks parent/children relationships.
+ * Auto-terminates after 5 minutes of inactivity.
+ */
+export async function notionDeletionCrawlWorkflow({
+  connectorId,
+  resourceQueue = [],
+}: {
+  connectorId: ModelId;
+  resourceQueue?: NotionDeletionCrawlSignal[];
+}) {
+  const seen = new Set<string>();
+  const topLevelWorkflowId = workflowInfo().workflowId;
+
+  // Clear workflow cache at start
+  await clearWorkflowCache({
+    connectorId,
+    topLevelWorkflowId,
+  });
+
+  // Helper to create unique key for seen tracking
+  const resourceKey = (resourceId: string, resourceType: string) =>
+    `${resourceType}:${resourceId}`;
+
+  // Set up signal handler
+  setHandler(notionDeletionCrawlSignal, (signal: NotionDeletionCrawlSignal) => {
+    const key = resourceKey(signal.resourceId, signal.resourceType);
+    // Only add to queue if not already seen
+    if (!seen.has(key)) {
+      resourceQueue.push(signal);
+    }
+  });
+
+  for (;;) {
+    // Wait for a resource signal, but stop the workflow if no signals arrive for 5 minutes
+    if (!(await condition(() => resourceQueue.length > 0, "5 minutes"))) {
+      return;
+    }
+
```

Comment:
Should seen() be reset before every while loop? Or am I confused?

Reply:
Having played with it, it definitely needs to be cleared, but that needs to happen at the end of the loop, not the beginning. Otherwise, a page that was moved out, and the moved back in will end up being undeletable because we'll ignore it in:
```ts
    if (!seen.has(key)) {
      resourceQueue.push(signal);
    }
```